tutorials/UsingJuMP.qmd

```diff
@@ -0,0 +1,101 @@
+---
+title: "Using Manopt.jl from within JuMP"
+author: Ronny Bergmann
+---
+
+In this tutorial we aim to illustrate how to use manifolds and the algorithms from [`Manopt.jl`](https://manoptjl.org) within the [JuMP](https://jump.dev) framework by implementing the [get started](getstarted.md) tutorial again using JuMP combining it with for example their [get started](https://jump.dev/JuMP.jl/stable/tutorials/getting_started/getting_started_with_JuMP/#Getting-started-with-JuMP) tutorial.
+
+```{julia}
+#| echo: false
+#| code-fold: true
+#| output: false
+using Pkg;
+cd(@__DIR__)
+Pkg.activate("."); # for reproducibility use the local tutorial environment.
+```
+
+```{julia}
+using JuMP, Manopt, Manifolds
+```
+
+First we generate the same data for the [Riemannian center of mass]() as before
+
+```{julia}
+n = 100
+σ = π / 8
+M = Sphere(2)
+p = 1 / sqrt(2) * [1.0, 0.0, 1.0]
+data = [exp(M, p,  σ * rand(M; vector_at=p)) for i in 1:n];
+```
+
+and we start by stating that our [JuMP model](https://jump.dev/JuMP.jl/stable/manual/models/) is
+
+```{julia}
+model = Model(Manopt.JuMP_Optimizer)
+```
+
+Next we add a [JuMP variable](https://jump.dev/JuMP.jl/stable/manual/variables/),
+where we specify the manifold using the `in` keyword, `start=` cares for the initialisation
+
+```{julia}
+@variable(model, p[i=1:3] in M, start=p[i])
+```
+
+For now this is restricted to array-type-representations of points and tangent vectors.
+
+## Variant I: Specify the
+
+One can specify a cost in the embedding and get an approximate gradient in the embedding that is “converted” into a Riemannian one, see [AD](AutomaticDifferentiation.md).
+
+```{julia}
+f(p) = sum(1 / (2 * n) * norm.(Ref(p) - data) .^ 2)
+grad_f(p) = sum(1 / n * grad_distance.(Ref(M), data, Ref(p)));
+# TODO: how to set the objective to an implemented function ?
+# TODO 100% unsure how to set a gradient since that seems to be nowhere documented ?!
+@objective(model, f)
```

Comment:
Yo me, two things are completely unclear even after reading a lot about JuMP:

1. If I have a cost function like f in line 51 – how do I set that as the objective?
2. If I have a gradient in the embedding how do I set a gradient in JuMP?

JuMP Seems to have super many magical commands to set stuff, but not explicit cost/grad things? @blegat

Reply:
You should do `@objective(model, Min, Manopt.ManifoldGradientObjective(f, grad_f))` but it needs https://github.com/JuliaManifolds/Manopt.jl/pull/455/

Reply:
Ah ok, I thought there was an alternative (the old?) way as well, but in the old code I never found how to provide a gradient.

Sure for the new way we need that PR as well.

Reply:
Ah sorry, I missed the context, this is the way:
https://jump.dev/JuMP.jl/stable/moi/submodules/Nonlinear/overview/#User-defined-operators

Reply:
Ah nice, then I'll try that one as well – would be nice to have both variants working here in the tutorial :)

Reply:
I agree

Reply:
If the Euclidean objective as a simple algebraic form, like `sum(x)`, you can also just use `@objective(model, Min, sum(x))` and not have to defined the gradient yourself